packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.tsx

```diff
@@ -0,0 +1,581 @@
+
+    if (newSelectedIndex !== undefined &&
+      newSelectedIndex !== null &&
+      newSelectedIndex !== this.state.selectedIndex) {
+      this.setState({
+        selectedIndex: newSelectedIndex
+      });
+    }
+  }
+
+  public render() {
+    let { swatchColorPickerButtonProps } = this.props;
+
+    // If we got button props, put the swatch color picker behind a button, otherwise
+    // render all of the items
+    let renderElement = swatchColorPickerButtonProps !== undefined ? this._buttonToRender() : this._fullSwatchColorPickerToRender();
+
+    return (
+      <div
+        ref={ this._resolveRef('_buttonWrapper') }
+        className={ css('ms-swatchColorPickerWrapper', styles.wrapper) }>
+        { renderElement }
+      </div>);
+  }
+
+  /**
+   * Get the selected item's index
+   * @param items - The items to search
+   * @param selectedId - The selected item's id to find
+   * @returns {number} - The index of the selected item's id, -1 if there was no match
+   */
+  private _getSelectedIndex(items: ISwatchColorPickerItemProps[], selectedId: string): number {
+    return findIndex(items, (item => (item.id === selectedId)));
+  }
+
+  /**
+   * Render the swatch color picker with a button in the given (collapased vs expanded state)
+   * @returns {JSX.Element[]} - an array of the elements to render
+   */
+  private _buttonToRender(): JSX.Element {
+    let { selectedIndex, expanded } = this.state;
+    let {
+      swatchColorPickerButtonProps,
+      disabled
+    } = this.props;
+
+    let colorToSet = this._getSelectedColorToSet();
+
+    return (
+      <div>
+        <DefaultButton
+          { ...swatchColorPickerButtonProps }
+          style={ { color: colorToSet && colorToSet } }
+          className={
+            css('ms-swatchColorPickerButton',
+              {
+                'is-expanded': expanded
+              }
+            )
+          }
+          onClick={ this._onButtonClick }
+          aria-haspopup={ true }
+          aria-expanded={ !disabled && expanded }
+          disabled={ disabled }
+          menuIconProps={ swatchColorPickerButtonProps.menuIconProps ?
+            swatchColorPickerButtonProps.menuIconProps :
+            { iconName: 'chevronDown' } }
+        />
+        { (!disabled && expanded) &&
+          this._onRenderContainer()
+        }
+      </div>
+    );
+  }
+
+  /**
+   * Gets the color for the selected index
+   * @returns {string} - The color for the selected index,
+   *   or null if: we are not updating the button icon with color,
+   *   there is not a valid selected index, or if we do not have a valid color
+   */
+  private _getSelectedColorToSet(): string {
+    let { selectedIndex } = this.state;
+    let {
+      swatchColorPickerItems,
+      updateButtonIconWithColor
+    } = this.props;
+
+    // Do we need to update the button with the selected
+    // item's color? If so, attempt to grab the color
+    if (updateButtonIconWithColor &&
+      selectedIndex &&
```

Comment:
It could be. I know that it gets set to -1 when the user executes against an already selected item (e.g. we need to make it no longer selected).

I think making assumptions is not a good idea, without the checks badness could result